### ISSUE TITLE
Research: Passkey-to-DID identity bridge architecture

### ISSUE THREAD

**Issue by jackzampolin** [jackzampolin/defradb.rs#475]
## Context

We want passkeys as the browser-side identity for DefraDB WASM. When the app starts, the user does a passkey prompt and gets identified. However, there are fundamental constraints:

### Key Constraints

1. **Passkey private keys are non-extractable** — you cannot use a passkey to sign a JWT or arbitrary data. WebAuthn only signs its own internal structure (authenticatorData || SHA-256(clientDataJSON)).

2. **Algorithm mismatch** — WebAuthn's default is ES256 (P-256/secp256r1). DefraDB identity supports Ed25519 and secp256k1, but NOT secp256r1.

3. **Ed25519 not available on platform authenticators** — TouchID, Windows Hello, and Android biometric only support ES256. Ed25519 is only on YubiKey 5.2+ hardware keys.

4. **Passkeys can authenticate but cannot be the DID** — the passkey gates access to the identity, it doesn't become the identity.

### Architecture Options

**Option A: Client-side key vault (edge-first)** — Recommended
- JS wrapper handles WebAuthn ceremony
- On first registration: generate Ed25519 keypair via Web Crypto API, encrypt with key derived from WebAuthn credential ID, store in IndexedDB
- On login: passkey authenticates → decrypt stored Ed25519 key → sign DID JWT → pass to DefraClient
- Pro: No server needed for identity. True edge-first.
- Con: Key recovery depends on browser storage persistence. Need backup/export.

**Option B: Server-issued identity**
- JS wrapper handles WebAuthn ceremony, sends assertion to server
- Server verifies assertion, looks up or creates DID identity mapped to credential
- Server issues time-limited DID JWT signed with server's key
- Pro: Standard pattern, easy recovery. Con: Server dependency for identity.

**Option C: Add secp256r1 (P-256) support to DefraDB identity**
- `Secp256r1PublicKey` already exists in `crates/crypto/` for verification
- Add P-256 JWT signing (ES256 algorithm) to `crates/identity/`
- Doesn't solve non-extractability, but enables P-256 identities from other sources
- Useful prep work regardless of A vs B

### The JS Wrapper

Since passkey operations are browser-platform APIs (navigator.credentials), this needs a JavaScript layer between the browser app and the WASM DefraClient. The WASM module stays pure Rust. The JS wrapper handles:
- WebAuthn registration/authentication ceremonies
- Encrypted key storage in IndexedDB
- JWT generation
- Passing identity into DefraClient

## Questions to Answer

1. For Option A: What's the encryption scheme for the stored keypair? HKDF from credential.rawId + salt → AES-GCM key?
2. For Option A: How does key backup/recovery work? Export encrypted blob? Multi-device passkey sync (iCloud/Google) means the same passkey works on multiple devices, but the IndexedDB keypair is device-local.
3. Should we support Option A AND Option B? (edge-first by default, server-issued as enterprise option)
4. Should we do Option C regardless? (small scope, unblocks other use cases)
5. What does the JS wrapper package look like? Separate npm package? Part of the wasm-pack output?

## Acceptance Criteria

- Architecture decision documented
- Specify JS wrapper API surface
- Identify what changes (if any) are needed in Rust crates
- Identify the key recovery / multi-device story

**Comment by jackzampolin** [comment]
## Update: Hublet Discovery

The [hublet](https://github.com/sourcenetwork/hublet) repo (SourceHub JS Wallet) already implements the core browser identity pattern we need:

### What Hublet Does
- **P-256 key generation** via Web Crypto API (`window.crypto.subtle.generateKey` with `extractable: false`)
- **IndexedDB storage** for `CryptoKey` objects (structured clone, non-extractable)
- **`did:key:` DID generation** from P-256 pubkeys using multicodec `0x1200`
- **Transaction signing** with P-256 ECDSA + low-S normalization
- **Session key derivation** — deterministic P-256 key from `SHA256(signature || message || pubkey) mod curve_order`

### Reframed Architecture

The browser identity should be a **P-256 keypair generated via Web Crypto API**, stored in IndexedDB. This is what hublet already does for SourceHub. The passkey becomes an optional UX gate, not the identity itself.

**Flow:**
1. App starts → check IndexedDB for existing keypair
2. If none: generate P-256 key via Web Crypto API → store in IndexedDB
3. P-256 public key → `did:key:` DID (same format as hublet)
4. Sign a JWT with ES256 (P-256) → pass to DefraClient via `set_identity(jwt)`
5. Optional: gate signing behind a passkey ceremony for UX polish

**What this means:**
- **P-256 support in DefraDB identity crate is now critical path** (see #478)
- The JS wrapper can extend/import hublet and add JWT generation
- SourceHub DID compatibility comes for free (same `did:key:` format)
- The "passkey signs DID JWT" vision works — just indirectly: passkey authenticates → unlock P-256 key from IndexedDB → P-256 key signs the JWT

**Remaining gaps vs hublet:**
- JWT signing/verification (ES256 algorithm) — hublet signs Cosmos txs, not JWTs
- Passkey ceremony integration (optional UX layer)
- Key backup/export/multi-device story

**Comment by jackzampolin** [comment]
## Update: Production Architecture — WebAuthn PRF Extension

### Discovery

The WebAuthn **PRF (Pseudo-Random Function) extension** solves the multi-device key derivation problem. It allows the authenticator to compute HMAC-SHA-256 using a per-credential internal secret and a relying-party-provided salt. The output is:

- **Deterministic**: same credential + same salt = same 32-byte output, every time
- **Cross-device**: synced passkeys (iCloud Keychain, Google Password Manager) produce identical output on all devices
- **Hardware-backed**: secret lives in secure enclave, never exposed to JavaScript
- **Origin-bound**: different origins get different output (domain separation built in)

This is the same pattern used by **Bitwarden** (vault decryption) and **1Password** (data encryption with passkeys) in production.

### Architecture

```
Passkey Auth + PRF Extension
       │
       ▼
PRF Output (32 bytes, deterministic)
       │
       ▼
HKDF-SHA-256(ikm=prfOutput, salt=appSalt, info="defra-p256-identity-v1")
       │
       ▼
P-256 scalar (reduce mod curve order n)
       │
       ▼
P-256 Private Key
       │
       ├──► did:key: DID (multicodec 0x1200, same format as hublet)
       ├──► ES256 JWT → DefraClient.set_identity(jwt)
       └──► SourceHub transaction signing (via hublet)
```

**No key storage needed.** The P-256 key is re-derived from the passkey on every authentication. Nothing persists in IndexedDB on the primary path.

### Two-Tier Strategy

| Tier | When | How |
|------|------|-----|
| **Primary** (PRF available) | Chrome 116+, Safari 18+, Edge 116+, Firefox 139+ | Deterministic key derivation from PRF output. No storage. |
| **Fallback** (PRF unavailable) | Windows Hello, Firefox/Android, older browsers | Generate random P-256 key, store in IndexedDB, gate behind passkey auth. Multi-device via export/import. |

Both tiers produce a P-256 keypair. Everything downstream (DID, JWT, DefraClient, sync) is identical.

### Browser Support (early 2026)

| Browser | PRF | Notes |
|---------|-----|-------|
| Chrome 132+ | ✅ | iCloud Keychain on macOS |
| Chrome/Android 116+ | ✅ | Google Password Manager |
| Safari 18+ | ✅ | iCloud Keychain only (not security keys) |
| Edge 116+ | ✅ | Chromium-based |
| Firefox 139+ | ✅ | Platform auth (NOT Android) |
| Windows Hello | ❌ | Doesn't support hmac-secret |
| Firefox/Android | ❌ | Doesn't support Google Password Manager PRF |

### Production Considerations

1. **Key loss = passkey deletion.** Different credentials produce different PRF outputs. If user deletes their only passkey, the identity is irrecoverable. Mitigation: support registering backup passkeys, or implement server-side encrypted key escrow.

2. **Key rotation via `second` salt.** The PRF extension supports evaluating two salts simultaneously. Use this for key rotation: derive both old and new keys, re-sign/re-encrypt, then switch.

3. **iOS 18.0–18.3 bug.** Apple fixed a data-loss bug in iOS 18.4. Target 18.4+ as minimum for PRF.

4. **Origin scoping.** Same passkey on different origins produces different keys. Identity is origin-bound. This is a security feature, not a bug.

### Dependencies

- **#478**: P-256 signing support in Rust identity crate (ES256 JWT verification on server)
- **JS wrapper**: Implements the WebAuthn ceremony + PRF + HKDF + JWT signing
- **hublet**: Reference implementation for P-256 key management and DID generation

**Comment by jackzampolin** [comment]
## Architecture Decision: Service Account Delegation Model

After discussion, the architecture is clearer than the original three options suggested. The key insight: **passkeys are local authentication, not identity. Identity lives in Orbis. DefraDB sees service accounts.**

### Three-Layer Separation

| Layer | What | Where |
|-------|------|-------|
| **Identity** | DID provisioned by SourceHub, private keys held in Orbis (threshold custody / DKG) | Orbis |
| **Service Account** | A delegated signing key authorized by Orbis to act on behalf of a DID. Each device gets its own. | Browser (IndexedDB) |
| **Authentication** | Passkey. Gates local access to the service account key. Signs nothing DefraDB cares about. | Browser (WebAuthn) |

### How It Works

**New user:**
1. User visits app, triggers passkey creation (WebAuthn ceremony)
2. JS wrapper generates a service account keypair (Ed25519), encrypts it under the passkey, stores in IndexedDB
3. Service account registers with Orbis → gets a delegation credential ("this key can act as this DID")
4. SourceHub provisions the DID as part of onboarding
5. DefraDB WASM starts with the service account as its identity

**Existing user, new device:**
1. Passkey syncs via iCloud/Google → WebAuthn ceremony on new device
2. New service account keypair generated on this device
3. Existing service account on another device authorizes the new one through Orbis
4. DefraDB starts, syncs data via P2P

**Revocation:** Lose a device → revoke that service account through Orbis using any other authorized service account. Each device is independently revocable.

### Browser DefraDB Runtime

Browser DefraDB has two external connections:

- **SourceHub** — ACP policy checks. "Can this DID write this document? Can it read that one?"
- **Other DefraDB nodes** — P2P replication, gated by ACP policies from SourceHub

Orbis is behind SourceHub, not something DefraDB talks to directly. The delegation chain (Orbis identity → service account) is something SourceHub understands.

**Mutation flow:**
1. User action → document update
2. Browser DefraDB signs it with the service account key
3. P2P syncs to server node
4. Server checks SourceHub ACP: "does this DID have write access per the policy on this collection?"
5. Accept or reject

**Read flow:** Same in reverse — P2P request comes in, ACP gates the response.

### Concrete Example: Game with Private Save Data

A game (Rust/WASM) uses DefraDB as its backend. Each user has their own encrypted save data, backed up to the developer's server via P2P.

- Developer defines ACP policies on SourceHub: "players read/write own saves, read shared world state, only admins write world events"
- Developer controls the **policy**, not the **data**
- Server replicates encrypted blobs it cannot decrypt
- Users own their data, gated by their Orbis identity

### Boundaries

- **JS wrapper** — handles passkey, IndexedDB, service account lifecycle, Orbis communication. Orchestration layer.
- **DefraDB WASM** — receives a service account key + delegation credential at startup. No WebAuthn awareness, no Orbis protocol awareness. Pure data engine.
- **SourceHub** — ACP policy oracle. DefraDB asks "is this DID allowed?" and SourceHub answers.

### What This Means for defradb.rs

Priority stack for browser support:

1. **WASM DefraClient accepts a delegated identity** — service account key + credential, verifiable without talking to Orbis
2. **SourceHub ACP works from WASM** — browser node can check policies
3. **P2P browser ↔ server** — replication channel (#472, #474)
4. **JS wrapper package** — passkey ceremony, service account management, hands identity to DefraDB

### Resolution of Original Options

- **Option A (client-side key vault):** ✅ Yes, but the key is a service account delegated by Orbis, not a standalone DID
- **Option B (server-issued):** Not needed — Orbis delegation replaces server-issued identity
- **Option C (P-256 support):** ✅ Already done (#491). Useful for P-256 identities from other sources.

**Comment by jackzampolin** [comment]
## Correction: Standalone DIDs Are First-Class

Important refinement to the architecture above. DefraDB's identity model is **device-first, always**:

- **Standalone mode (default):** Device generates a DID. That IS the identity. Hardware is the root of trust. No Orbis, no SourceHub required. Works everywhere — browser, server, IoT, embedded, edge.
- **Orbis mode (optional overlay):** Orbis manages identity with threshold custody, device gets a delegated service account. For multi-device, enterprise, browser-with-backup scenarios.

**DefraDB doesn't care which mode it's in.** It sees a signing key. In standalone mode the key is self-rooted (the DID IS the key). In Orbis mode it's a delegation. Same interface.

### Core Design Principle

Every operation is cryptographically signed. Every document has provenance. The hardware is the default trust anchor.

A sensor with a TPM signs its writes the same way a browser with a passkey-gated key does. An IoT device uses whatever secure element it has. A server uses its keyring. A browser uses a JS-generated key protected by a passkey. All produce the same thing: a signed operation from a DID.

This is DefraDB's intended architecture: **cryptographic proof for all operations in the system, device-rooted identity as the base case, Orbis delegation as an optional enhancement.**

### Updated Priority

1. Device-generated standalone DIDs work everywhere DefraDB runs (including WASM) — this is the base case
2. The JS wrapper in browser handles passkey → local key protection (UX layer, not identity layer)
3. Orbis delegation is an optional mode that layers on top without changing DefraDB's identity interface
4. SourceHub ACP works regardless of standalone vs Orbis mode

**Comment by jackzampolin** [comment]
Consolidating the WASM browser-client track under #474 as the single umbrella. Closing this as a separate issue; comments here are preserved on the issue history. Reopen if/when this sub-concern needs its own tracking.
